File: 2021/09/30/main.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        s, n, mx = 0, len(nums), nums[0]
        for v in nums:
            s += v
            mx = max(mx, v)

        if s%k != 0 or s < mx*k:
            return False

        nums.sort(reverse=True)

        u = [False] * n

        def dfs(u, k, j, s, t):
            if k == 1:
                return True

            if s == t:
                return dfs(u, k-1, 0, 0, t)

            for i in range(j, n):
                if not u[i] and s+nums[i] <= t:
                    u[i] = True
                    if dfs(u, k, i+1, s+nums[i], t):
                        return True
                    u[i] = False

            return False

        return dfs(u, k, 0, 0, s//k)

File: 2021/09/30/main.go
package main

import "sort"

func canPartitionKSubsets(nums []int, k int) bool {
	s, n := 0, len(nums)
	mx := nums[0]
	for _, v := range nums {
		s += v
		if mx < v {
			mx = v
		}
	}

	if s%k != 0 || s < mx*k {
		return false
	}

	sort.Sort(sort.Reverse(sort.IntSlice(nums)))

	u := make([]bool, n)

	return dfs(nums, u, k, 0, 0, s/k)
}

func dfs(nums []int, u []bool, k, j, s, t int) bool {
	if k == 1 {
		return true
	}

	if s == t {
		return dfs(nums, u, k-1, 0, 0, t)
	}

	for i := j; i < len(nums); i++ {
		if !u[i] && s+nums[i] <= t {
			u[i] = true
			if dfs(nums, u, k, i+1, s+nums[i], t) {
				return true
			}
			u[i] = false
		}
	}

	return false
}
